crates/sui-adapter/src/adapter.rs

```diff
@@ -1345,15 +1365,36 @@ fn missing_unwrapped_msg(id: &ObjectID) -> String {
     )
 }
 
-fn convert_type_argument_error(idx: usize, error: VMError) -> ExecutionError {
+fn convert_type_argument_error(
+    idx: usize,
+    error: VMError,
+    vm_resolver: impl Fn(&ModuleId, FunctionDefinitionIndex) -> Option<Identifier>,
+) -> ExecutionError {
     use move_core_types::vm_status::StatusCode;
     use sui_types::messages::EntryTypeArgumentErrorKind;
     let kind = match error.major_status() {
         StatusCode::LINKER_ERROR => EntryTypeArgumentErrorKind::ModuleNotFound,
         StatusCode::TYPE_RESOLUTION_FAILURE => EntryTypeArgumentErrorKind::TypeNotFound,
         StatusCode::NUMBER_OF_TYPE_ARGUMENTS_MISMATCH => EntryTypeArgumentErrorKind::ArityMismatch,
         StatusCode::CONSTRAINT_NOT_SATISFIED => EntryTypeArgumentErrorKind::ConstraintNotSatisfied,
-        _ => return error.into(),
+        _ => return convert_vm_error(error, vm_resolver),
     };
     ExecutionErrorKind::entry_type_argument_error(idx as TypeParameterIndex, kind).into()
 }
+
+pub fn with_vm_resolver<
```

Comment:
Is it fair to say that the only kind of lambda that gets passed to `convert_vm_error` is one that comes from calling this function, and if so, does it make sense for `convert_vm_error` to just accept the `vm` and `state_view` directly, and, in effect inline the call to the lambda?

Reply:
I thought about this too. I think the trouble comes when we want to use `convert_vm_error` when we _don't_ have a handle on `vm` or `state_view` (in which case, it's handy to have a resolver function to do things). 

It might be that the intent behind `convert_vm_error` is to say "hey yeah we should always have a valid handle to vm for these errors. _However_ prior to this PR, there definitely was a case where we didn't have a handle on `vm`, and used `convert_vm_error`, so the intent here was maybe more aspirational than anything else. So yes, _now_ that I've shifted things around, it's become obvious that it's possible to invoke `convert_vm_error` with a valid `vm` always. 

So: Now I'm OK with passing `vm` and `state_view` directly. I will wait for @tnowacki to weigh in on this, versus passing a lambda.

Reply:
More details on how we resolved errors and didn't have a handle on `vm`: https://github.com/MystenLabs/sui/pull/7306/files#r1087087209

Essentially, the `convert_vm_error` function is a bit _too_ much of a catch-all error resolution function, so teasing a part the actual state/errors possible at various points during execution is going to be the thing that unlocks better errors. Similar story with the location info issue in `id_leak_verifier.rs` really.

Reply:
But we could just shuffle around publish so we have the VM instance? there isn't really a reason not to

Reply:
works for me, done!